sys/ztimer: add ztimer64_xtimer_compat complete xtimer replace module

fjmolinas opened the pull request:
### Contribution description

This PR makes `ztimer_xtimer_compat` use `ztimer64_usec` implementing all `xtimer` api, with this `ztimer64_usec` can be used as a drop-in replacement for `xtimer`.

### Testing procedure

- Green Murdock should be enough, #17365 will flag possible remaining offenders.

### Issues/PRs references

Split from #17365

fjmolinas commented:
@leandrolanzieri do you mind taking a quick look I changed some of the xtimer/ztimer modelings, and I want to be sure I'm not re-introducing issues.

leandrolanzieri commented on sys/xtimer/Kconfig:
```diff
@@ -9,9 +9,7 @@ menuconfig MODULE_XTIMER
-
-    # use timer peripheral unless ztimer compatibility module is used
-    select MODULE_PERIPH_TIMER if HAS_PERIPH_TIMER && !MODULE_XTIMER_ON_ZTIMER && !MODULE_ZTIMER_XTIMER_COMPAT
+    select MODULE_PERIPH_TIMER if HAS_PERIPH_TIMER
```
I think the idea of this was not to force the timer backend when compatibility is in place, leaving the option of the hardware to the Ztimer config. Why does this change?

fjmolinas replied:
will split out in another commit if justfied

fjmolinas replied:
Hmm well all green

fjmolinas replied:
@leandrolanzieri does everything else Kconfig related look good to you?

fjmolinas commented:
@kaspar030 could you give this one a look? IMO its missing only a rebase, squash, uncrustify the header file and this https://github.com/RIOT-OS/RIOT/pull/17670#discussion_r811124744

fjmolinas commented:
Maybe @kfessel wants to take a look as well

kaspar030 commented on sys/net/dsm/Makefile.dep:
```diff
@@ -1 +0,0 @@
-FEATURES_REQUIRED += sock_dtls
```
why is this deleted?

fjmolinas replied:
Two reasons:

- its not a feature but a Module
- that Makefile.dep was never included

I just moved the dependency to 9b856c8e1429e89925ded0ed9786900546a2bc21, I can split this one, its one of those things that just showed up during the ci tests...

fjmolinas replied:
Should I split it out?

kaspar030 replied:
I guess yeah, it is completely unrelated. you can expect a quick ack from me :)

kaspar030 commented:
> @kaspar030 could you give this one a look? IMO its missing only a rebase, squash, uncrustify the header file and this https://github.com/RIOT-OS/RIOT/pull/17670#discussion_r811124744

LGTM. Can't say much about the Kconfig, rest looks fine. (apart from that spurious file delete)

kaspar030 commented on sys/xtimer/Kconfig:
```diff
@@ -9,7 +9,8 @@ menuconfig MODULE_XTIMER
     depends on HAS_PERIPH_TIMER
     depends on TEST_KCONFIG
-    select MODULE_PERIPH_TIMER if HAS_PERIPH_TIMER
+    # select MODULE_PERIPH_TIMER if HAS_PERIPH_TIMER
```
@leandrolanzieri

leandrolanzieri replied:
The comment? yes, that means you still keep
```
    select MODULE_PERIPH_TIMER if HAS_PERIPH_TIMER && !MODULE_XTIMER_ON_ZTIMER && !MODULE_ZTIMER_XTIMER_COMPAT
```

right?

fjmolinas replied:
Yes, I meant removing the comments, ti was a note for myself when rebasing.

leandrolanzieri commented on sys/xtimer/Kconfig:
```diff
@@ -9,8 +9,7 @@ menuconfig MODULE_XTIMER
     depends on HAS_PERIPH_TIMER
     depends on TEST_KCONFIG
-
-    # use timer peripheral unless ztimer compatibility module is used
```
Let's keep the comment as well

fjmolinas commented:
Thanks @kaspar030 and @leandrolanzieri I'll then wait for https://github.com/RIOT-OS/RIOT/pull/17685 and rebase.

kfessel commented on sys/event/timeout.c:
```diff
@@ -10,7 +10,7 @@
 #include "kernel_defines.h"
 #include "event/timeout.h"
-#if IS_USED(MODULE_ZTIMER_USEC)
+#if IS_USED(MODULE_ZTIMER64_USEC) && !IS_USED(MODULE_XTIMER)
```
this forces ztimer64_usec for every thing that uses event time out (even though event time out clearly just needs ztimer usec)

kaspar030 replied:
... unless MODULE_EVENT_TIMEOUT_ZTIMER is used.

The "old" event_timeout depends on xtimer semantics. users of that block low power mode, whether they use xtimer_compat based on ztimer or ztimer64 doesn't matter. pulling the clock under them breaks them in any case.

xtimer_compat doesn't solve the lack of power management awareness of xtimer-using code, it just allows using ztimer as base timer.

kfessel replied:
if something is using just the event timeout interface it does not depend on xtimer semantics -> only if it uses xtimer it implies xtimer semantics
in makefile.dep there is one (direct) user event timeout "asymcute" that does not other wise specify a timer implementation and it does not use any xtimer

kaspar030 replied:
now I see your point I think. There's no reason to use a ztimer64 unless ztimer_xtimer_compat is used, if that needs one.
But in that case, event timeout with using the compat header is not compatible anymore with event timeout just using ztimer32.

Can't we just remove xtimer as backend timer here? keep the old api (without specifying a clock -> ztimer_usec is used). (out of scope I guess).

kfessel replied:
in ecence event timeout dos not depend on xtimer i am not sure if this should be kept working on an non xtimer configuration but using xtimer_compat64 for its implementation is realy super plus ungood

kaspar030 replied:
If no ztimer-generation code is using event_timer, why is compat64 worse compared to compat32?

(*if* ztimer generation code is using event_timer, it is using ztimer_usec(32) anyways)

kfessel replied:
all users of event timeout are either ztimer(32) enabled or use no other timer (asymcute) -> using ztimer64 is ++ungood
<details>
I see i need to get better at getting my thoughts into written form
</details>

kfessel reviewed (changes requested):
Please keep old  xtimer_compat and add xtimer_compat64.h (true compatibility)

xtimer_compat64 can be auto selected xtimer_compat can be used on a per file basis (as we did with event/timeout.c)

kaspar030 commented:
> Please keep old xtimer_compat and add xtimer_compat64.h (true compatibility)

can you elaborate? you want to keep two versions, with one only implementing half of the api?

fjmolinas commented:
> Please keep old xtimer_compat and add xtimer_compat64.h (true compatibility)

Why?

kfessel commented:
@kaspar030 : yes  please lets keep two versions one just half API at just a quater of the cost in every aspekt of its use 
There also isn't much API (one header and it does not do very much -> no huge maintainace burden
But many things can be made ztimer compatible by just adding the three include lines off event/timout.c on top (I mean ztimer without ztimer64)

fjmolinas commented:
> xtimer_compat64 can be auto selected xtimer_compat can be used on a per file basis (as we did with event/timeout.c)

Is this really necessary? It was quite a pain to model the dependency already, this is adding another layer to code that we want to remove anyway... The potential users of these inline are going to be few, the `ztimer_xtimer_compat` users for `event_timeout` is also something that will be removed, all should just use the proper `event_timeout_ztimer_init` functions, it was just easier at the time to not have to migrate all users (there are not thaaaat many `event_timeout` users either)

kfessel commented:
> ztimer_xtimer_compat users for event_timeout is also something that will be removed 

hopefully not many of them need ztimer64_usec (but can deal with msec or ztimer_usec)

kfessel commented:
Why did one use xtimer prior to ztimer - there wasn't something else.
What is the second easyiest xtimer to ztimer conversion - `#include "ztimer/xtimer_compat.h"` and use `ztimer_usec` this will solve 80% of all conversions and most of thewm will work better than before. (the new one will solve 100% but at comparably huge cost)
"ztimer/xtimer_compat.h" is a nice header to ztimer_usec


FN: easiest would be to use module `ztimer_xtimer_compat` (which i never tried since the second easiest is so easy)

kaspar030 commented:
> easiest would be to use module ztimer_xtimer_compat (which i never tried since the second easiest is so easy)

You do realize that the whole point of this PR is to make this even easier?
Before, only a subset of the applications could be "converted" by using the compat module. Now, all xtimer applications can be used with it. In both cases (and with xtimer), the usec timer needs to stay active, blocking low power. But wirh this PR, the set of applications that need "xtimer_on_ztimer" (IMO, the worst case scenario) is reduced to 0.

I think I don't see the issue still @kfessel, can you maybe give more specific examples? Are we talking about a general issue or a specific implementation detail?

kfessel commented:
> > easiest would be to use module ztimer_xtimer_compat (which i never tried since the second easiest is so easy)
> 
1.
> You do realize that the whole point of this PR is to make this even easier? Before, only a subset of the applications could be "converted" by using the compat module. Now, all xtimer applications can be used with it. In both cases (and with xtimer), the usec timer needs to stay active, blocking low power. But wirh this PR, the set of applications that need "xtimer_on_ztimer" (IMO, the worst case scenario) is reduced to 0.
> 
2.
> I think I don't see the issue still @kfessel, can you maybe give more specific examples? Are we talking about a general issue or a specific implementation detail?

1.:
Yes i do realize that this provides a near 100% replacement for xtimer.
I think this PR i generally a good idea and i think that `USE_MODULE += xtimer xtimer_compat` should lead to xtimer_compat64 being used.

2.:
I generaly do think that ztimer64_usec should be avoided.

2.:
ztimer/xtimer_compat.h is an easy way to upgrade applications that does not need xtimer with 64 Bit
and if that application did not use xtimer_now  one can even enable pm after just replacing "#include xtimer.h" for "#include ztimer/xtimer_compat.h" and be done with that conversion

--> I think there should be a second header file that is a 100% xtimer replacement (and if someone just does USE_MODULE + =xtimer" it should be used, but we should keep that header that is a 80% xtimer replacement but works at no cost 
using ztimer64_usec is high cost not only for that extra clock but also the ztimer64 that replace the xtimer (still lower than it).

event_timeout is only an example that happend to be right in this PR, it contains a complete xtimer to ztimer conversion at no cost. Do you see that?

fjmolinas commented:
I added as asked for a `ztimer64/xtimer_compat.h" header, but I also removed `compat` usage from `event/timeout`, its honestly simply not worth the effort anymore to keep both. Let's see what Murdock thinks.

In a follow up what I'll do is enable `ztimer_xtimer_compat` by default, see which modules actually require the whole API, and have those explicitly select `ztimer64_xtimer_compat`.

fjmolinas commented:
> In a follow up what I'll do is enable ztimer_xtimer_compat by default, see which modules actually require the whole API, and have those explicitly select ztimer64_xtimer_compat.

So I'm modeling `ztimer64_xtimer_compat` as an "extension", both modules are pseudomodules anyway, but having it this way makes it easy to model in Make and Kconfig, since there will be no `ifeq` dependencies.

fjmolinas commented:
@kaspar030 @kfessel I'll split the event_timeout not supporting `ztimer` PR out.

kfessel commented:
somthing went wrong (event/timeout.c got added)

fjmolinas commented:
> somthing went wrong (event/timeout.c got added)

ups wrong rebase, re-added

kaspar030 reviewed (approved):
ACK.

kfessel commented on sys/include/ztimer/xtimer_compat.h:
```diff
@@ -256,7 +256,6 @@ static inline bool xtimer_less64(xtimer_ticks64_t a, xtimer_ticks64_t b)
    static inline xtimer_ticks64_t xtimer_ticks_from_usec64(uint64_t usec);
    static inline uint32_t xtimer_usec_from_ticks(xtimer_ticks32_t ticks);
    static inline uint64_t xtimer_usec_from_ticks64(xtimer_ticks64_t ticks);
-
```
please let [ztimer/xtimer_compat.h](https://github.com/RIOT-OS/RIOT/pull/17670/files#diff-0f4bb1037a184712d68add099b16b7a129a4fd4dfc91471f5026730078674f48) unmodified

kfessel commented on sys/include/ztimer64/xtimer_compat.h:
```diff
@@ -0,0 +1,313 @@
+    return ztimer64_now(ZTIMER64_USEC);
+}
+
+
```
please rmove that empty line
```suggestion
```

kfessel reviewed (approved):
two nitpicks,
i think you can squash them right in and skip the murdock build

fjmolinas commented:
Squashed in both nitpicks @kfessel

fjmolinas commented:
All green @kaspar030 @kfessel

fjmolinas commented:
Thanks, @kaspar030 @kfessel!
